aws/jsii
# fix(jsii-pacmak): computeIfAbsent throws ConcurrentModificationException

---

Java generated modules use cache to store the result of `resolveClass` which is called when ever `jsiiStaticGet`/`jsiiGet` is called. The cache is implemented as a  `HashMap`, to fill the cache we use `computeIfAbsent`:

```Java
this.cache.computeIfAbsent(MODULE_TYPES.get(fqn), this::findClass);
```
> [code](https://github.com/aws/jsii/blob/master/packages/jsii-pacmak/lib/targets/java.ts#L2533)

When either of the get methods is called on a type which in itself includes a type, the first get (the root type get) will call get on the inner type. When `computeIfAbsent` is called from the root get, the compute method will trigger another call to `computeIfAbsent` to resolve the inner type, since all types share the same cache it means that the inner `computeIfAbsent` call will modify the cache while the root `computeIfAbsent` is not yet completed, resulting it a `ConcurrentModificationException`. 

This effects all Java >= 9 applications as a change to the behavior of `computeIfAbsent` was introduced in Java 9.  From the [Javadocs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/HashMap.html#computeIfAbsent(K,java.util.function.Function)):

```
Throws:
ConcurrentModificationException - if it is detected that the mapping function modified this map
```
The use of  `computeIfAbsent `  was introduced in https://github.com/aws/jsii/pull/1605 and released in `v1.4.1`. 


This PR replaces `computeIfAbsent` with explicitly checking if the value exists in the cache and if not adding it (essentially old style `computeIfAbsent`).

### Test
Steps to verify fix:

1. Build and package a sample module (see repro) using a local version of `jsii` and `jsii-pacmak`.
2. Add the generated jar to a local `mvn` repo.
3. Added the package as a dependency to a Java11 application.
4. Verify the exception is no longer thrown. 

### Repro

A repro can be found in this [repo](https://github.com/NetaNir/jsii-nested-get-repro)


By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0
